### ISSUE TITLE
Newlib is not able to resolve patches directory correctly if newlib-nano companion library is enabled

### ISSUE THREAD

**Issue by nolange** [crosstool-ng/crosstool-ng#1973]
If newlib-nano is enabled then the newlib-nano directory is searched for patches - but this directory does not contain the patches.
Correct place to search would be `/usr/share/crosstool-ng/packages/newlib/4.3.0.20230120`

(My own branch of crosstools-ng, would take some time for me to redo the steps on upstream):

```
[DEBUG]    ==> Return status 0
[DEBUG]    ==> Executing:  'touch' '/tmp/a/.build/src/.newlib-4.3.0.20230120.extracted'
[DEBUG]    ==> Return status 0
[DEBUG]    ==> Executing:  'rm' '-f' '/tmp/a/.build/src/.newlib-4.3.0.20230120.extracting'
[DEBUG]    ==> Return status 0
[EXTRA]    Patching newlib-4.3.0.20230120
[DEBUG]    ==> Executing:  'touch' '/tmp/a/.build/src/.newlib-4.3.0.20230120.patching'
[DEBUG]    ==> Return status 0
[DEBUG]    Entering '/tmp/a/.build/src/newlib-4.3.0.20230120'
[DEBUG]    Looking for patches in '/usr/share/crosstool-ng/packages/newlib-nano/4.3.0.20230120'...
[DEBUG]    Looking for patches in '/usr/share/crosstool-ng/packages/newlib-nano'...
```

**Comment by nolange** [comment]
with newlib-nano as companion:

```
[DEBUG]    Package iterator: override CT_MPC_SRC_DIR=/tmp/a/.build/src
[DEBUG]    Package iterator: set basename='newlib-4.3.0.20230120'
[DEBUG]    Package iterator: set pkg_name='newlib-nano'
[DEBUG]    Package iterator: set version='4.3.0.20230120'
[DEBUG]    Package iterator: set pkg_dir='newlib-nano/4.3.0.20230120'
[DEBUG]    Package iterator: set src_release='y'
[DEBUG]    Package iterator: set mirrors='https://sourceware.org/pub/newlib'
[DEBUG]    Package iterator: set archive_filename='newlib-@{version}'
[DEBUG]    Package iterator: set archive_dirname='newlib-@{version}'
[DEBUG]    Package iterator: set archive_formats='.tar.gz'
```

without:

```
[DEBUG]    Package iterator: override CT_GCC_SRC_DIR=/tmp/a/.build/src
[DEBUG]    Package iterator: set basename='newlib-4.3.0.20230120'
[DEBUG]    Package iterator: set pkg_name='newlib'
[DEBUG]    Package iterator: set version='4.3.0.20230120'
[DEBUG]    Package iterator: set pkg_dir='newlib/4.3.0.20230120'
[DEBUG]    Package iterator: set src_release='y'
[DEBUG]    Package iterator: set mirrors='https://sourceware.org/pub/newlib'
[DEBUG]    Package iterator: set archive_filename='@{pkg_name}-@{version}'
[DEBUG]    Package iterator: set archive_dirname='@{pkg_name}-@{version}'
[DEBUG]    Package iterator: set archive_formats='.tar.gz'
```

Later. `bundled_patch_dir="${CT_LIB_DIR}/packages/${pkg_dir}"` is set.

Seems the issue is, that the companion libs are downloaded before libc

**Comment by nolange** [comment]
reproducing defconfig:

``` ini
CT_CONFIG_VERSION="4"
CT_PREFIX_DIR="/tmp/arm"
# CT_PREFIX_DIR_RO is not set
# CT_LOG_PROGRESS_BAR is not set
CT_ARCH_ARM=y
CT_MULTILIB=y
# CT_ARCH_USE_MMU is not set
CT_TARGET_VENDOR="none"
CT_BINUTILS_PLUGINS=y
CT_BINUTILS_RELRO=y
CT_LIBC_NEWLIB_IO_C99FMT=y
CT_LIBC_NEWLIB_IO_LL=y
CT_LIBC_NEWLIB_IO_FLOAT=y
CT_LIBC_NEWLIB_DISABLE_SUPPLIED_SYSCALLS=y
CT_GCC_V_8=y
CT_CC_GCC_EXTRA_CONFIG_ARRAY="--with-zstd"
CT_CC_GCC_BUILD_ID=y
CT_CC_GCC_LNK_HASH_STYLE_GNU=y
CT_CC_LANG_CXX=y
CT_COMP_LIBS_NEWLIB_NANO=y
CT_NEWLIB_NANO_INSTALL_IN_TARGET=y
CT_LIBC_NEWLIB_NANO_IO_C99FMT=y
CT_LIBC_NEWLIB_NANO_DISABLE_SUPPLIED_SYSCALLS=y
```

**Comment by cpackham** [comment]
@nolange I just tried your defconfig above and the toolchain built fine for me.

**Comment by nolange** [comment]
Fails for me. I am not deep enough into `ct-ng`'s entrails to know, but is it possible that the list of downloads is generated by using a directory iterations. Meaning the order depends on whatever package was added last?

I reproduced it from scratch in a container, using `podman run --rm -it debian:bullseye`

Setup of ct-ng:

``` bash
apt-get update; apt-get install -y build-essential python3 git libzstd1 autoconf automake autotools-dev libncurses-dev gperf bison flex texinfo help2man gawk wget rsync bzip2 libtool-bin unzip

mkdir /build; cd build
git clone https://github.com/crosstool-ng/crosstool-ng.git
cd crosstool-ng

./bootstrap 
./configure
make && make install
cd /tmp; rm -rf /tmp/build

# move newlib after newlib-nano in the directory listing
# check find /usr/local/share/crosstool-ng/packages/ -maxdepth 1
mv /usr/local/share/crosstool-ng/packages/newlib /usr/local/share/crosstool-ng/packages/newlib_2
mv /usr/local/share/crosstool-ng/packages/newlib_2 /usr/local/share/crosstool-ng/packages/newlib

# Add user for build
useradd -m -s /bin/bash build

su - build
```

The ARM Build itself - as build user (just separated for clarity):

``` bash
mkdir /tmp/build
cd /tmp/build

cat >/tmp/arm.defconfig << EOF
CT_CONFIG_VERSION="4"
CT_PREFIX_DIR="/tmp/arm"
# CT_PREFIX_DIR_RO is not set
# CT_LOG_PROGRESS_BAR is not set
CT_ARCH_ARM=y
CT_MULTILIB=y
# CT_ARCH_USE_MMU is not set
CT_TARGET_VENDOR="none"
CT_BINUTILS_PLUGINS=y
CT_BINUTILS_RELRO=y
CT_LIBC_NEWLIB_IO_C99FMT=y
CT_LIBC_NEWLIB_IO_LL=y
CT_LIBC_NEWLIB_IO_FLOAT=y
CT_LIBC_NEWLIB_DISABLE_SUPPLIED_SYSCALLS=y
CT_GCC_V_8=y
CT_CC_GCC_EXTRA_CONFIG_ARRAY="--with-zstd"
CT_CC_GCC_BUILD_ID=y
CT_CC_GCC_LNK_HASH_STYLE_GNU=y
CT_CC_LANG_CXX=y
CT_COMP_LIBS_NEWLIB_NANO=y
CT_NEWLIB_NANO_INSTALL_IN_TARGET=y
CT_LIBC_NEWLIB_NANO_IO_C99FMT=y
CT_LIBC_NEWLIB_NANO_DISABLE_SUPPLIED_SYSCALLS=y
EOF

ct-ng DEFCONFIG=/tmp/arm.defconfig defconfig
ct-ng build
```

A first hint that something is wrong is already:

```
[EXTRA]    Retrieving 'newlib-4.3.0.20230120'
[WARN ]    Not verifying 'newlib-4.3.0.20230120.tar.gz': digest missing
```

There is probably a way to check way faster than a full build (attempt) to see if the patches
are applied. For example this should be true if newlib got patched:

``` bash
grep 'if __ARM_FEATURE_MVE' .build/src/newlib-4.3.0.20230120/newlib/libc/machine/arm/setjmp.S
```

**Comment by KirillSmirnov** [comment]
Should be fixed in #1998

**Comment by cpackham** [comment]
Fixed via #1998
